File: HR-js/disjoint-set/kruskal.py
class Node:
    def __init__(self, rank, data, parent):
        self.rank = rank
        self.data = data
        self.parent = parent

class DisjointSet:
    def __init__(self):
        self.map = dict()
    
    def union(self ,data1, data2):
        node1 = self.map[data1]
        node2 = self.map[data2]
        parent1 = self.findSet(node1)
        parent2 = self.findSet(node2)

        if parent1.data == parent2.data:
            return False
        
        if parent1.rank >= parent2.rank:
            parent1.rank = (parent1.rank + 1) if parent1.rank == parent2.rank else parent1.rank         
            parent2.parent = parent1
        else:
            parent1.parent = parent2
        return True
    
    def findSet(self, node):
        parent = node.parent
        if parent == None:
            return node
        node.parent = self.findSet(node.parent)
        return node.parent

    def makeSet(self, data):
        node = Node(0, data, None)
        self.map[data] = node


class Edge:
    def __init__(self, fr, to, weight):
        self.fr = fr
        self.to = to
        self.weight = weight


def kruskals(gNodes, gFrom, gTo, gWeight):
    edgeArr = list()
    for i in range(0, len(gFrom)):
        edge = Edge(gFrom[i], gTo[i], gWeight[i])
        edgeArr.append(edge)
    sortedEdgeArr = sorted(edgeArr, key=lambda edge: edge.weight)

    dSet = DisjointSet()
    for i in range(1, gNodes + 1):
        dSet.makeSet(i)

    edgeCount = 0
    i = 0
    total = 0
    while (edgeCount < gNodes - 1 and i < len(sortedEdgeArr)):
        edge = sortedEdgeArr[i]
        flag = dSet.union(edge.fr, edge.to)
        if flag:
            total = total + edge.weight
        i = i + 1
    return total

def main():
    gNodes = 4
    gFr = [1, 1, 4, 2, 3, 3]
    gTo = [2, 3, 1, 4, 2, 4]
    gWeight = [5, 3, 6, 7, 4, 5]

    total = kruskals(gNodes, gFr, gTo, gWeight)
    print(total)

main()
    

File: HR-js/disjoint-set/kruskal.js
class Node {
    constructor(rank, data, parent) {
        this.rank = rank;
        this.data = data;
        this.parent = parent;
    }
}

class DisjointSet {
    constructor() {
        this.map = new Map();
    }
    union(data1, data2) {
        const node1 = this.map.get(data1);
        const node2 = this.map.get(data2);
        const parent1 = this.findSet(node1);
        const parent2 = this.findSet(node2);

        //if they are part of same set do nothing
        if (parent1.data === parent2.data) {
            return false;
        }

        //else whoever's rank is higher becomes parent of other
        if (parent1.rank >= parent2.rank) {
            parent1.rank = (parent1.rank === parent2.rank) ? parent1.rank + 1 : parent1.rank;
            parent2.parent = parent1;
        } else {
            parent1.parent = parent2;
        }
        return true;
    }
    /**
     * Find the representative recursively and does path
     * compression as well.
     */
    findSet(node) {
        const parent = node.parent;
        if (parent === null) {
            return node;
        }
        node.parent = this.findSet(node.parent);
        return node.parent;
    }
    makeSet(data) {
        const node = new Node(0, data, null);
        this.map.set(data, node);
    }
}

class Edge {
    constructor(from, to, weight) {
        this.from = from;
        this.to = to;
        this.weight = weight;
    }
}

function kruskal(gNodes, gFrom, gTo, gWeight) {
    /* 
    sort edges based on weight
    create disjoint set 
    */
    const edgeArr = [];
    for (let i = 0; i < gFrom.length; i++) {
        const edge = new Edge(gFrom[i], gTo[i], gWeight[i]);
        edgeArr.push(edge);
    }
    const sortedEdgeArr = edgeArr.sort((edge1, edge2) => {
        return edge1.weight - edge2.weight;
    });
    const dSet = new DisjointSet();
    for (let i = 1; i <= gNodes; i++) {
        dSet.makeSet(i);
    }
    let edgeCount = 0;
    let i = 0;
    let total = 0;
    while (edgeCount < gNodes - 1 && i < sortedEdgeArr.length) {
        const edge = sortedEdgeArr[i];
        if (dSet.union(edge.from, edge.to)) {
            total += edge.weight;
        }
        i++;
    }
    return total;
}

function main() {
    const gNodes = 4;
    const gFrom = [1, 1, 4, 2, 3, 3];
    const gTo = [2, 3, 1, 4, 2, 4];
    const gWeight = [5, 3, 6, 7, 4, 5];

    const total = kruskal(gNodes, gFrom, gTo, gWeight);
    console.log(total);
}

main();


